### PR TITLE
[stubtest] Verify __all__ exists in stub

### LINKED ISSUE #13300: Should stubtest verify the presence of `__all__` in a stub?
See https://github.com/python/mypy/pull/12214

Currently, stubtest will not ensure that `__all__` is present in the stub if it is present at runtime.

We will always check whether the stub has the symbols we need, so ensuring that `__all__` is in the stub may just cause additional busywork for maintainers of stubs.

However, users may do `module.__all__` at runtime. If stubtest doesn't ensure that `__all__` is present in the stub, this means such users could get false positives when using such stubs.

However, if `__all__` isn't in `__all__` is it a public API? stubtest *will* ensure that `__all__` is present in the stub if `__all__` is in `__all__`.

Anyway, busywork is my main concern here. The change itself is very easy: https://github.com/python/mypy/pull/12214/commits/02b4f14eed85bf81d1dc8ce9af8452ea8476fb93

### PR DESCRIPTION
**Pull request by srittau**
Previously it wasn't an error if runtime included an
`__all__` declaration, but the stubs did not. This PR
changes this to reflect the consensus that it would
be a good idea to ensure consistency in this case.

Fixes #13300

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Does this cause a lot of new errors in typeshed?

**Comment by srittau** [comment]
How can we check that? If it does, I suggest we just mass add `.*\.__all__  # TODO` to the stubtest allowlists, like @Avasam suggested.

**Comment by JelleZijlstra** [comment]
I tried locally (by installing this PR branch and running typeshed tests), and didn't see any new errors—which actually seems a little suspicious.

**Comment by srittau** [comment]
> I tried locally (by installing this PR branch and running typeshed tests), and didn't see any new errors—which actually seems a little suspicious.

Very. I suspect there to be a quite significant number of hits.

**Comment by JelleZijlstra** [comment]
It was because of #18007; I forgot that stubtest doesn't do anything if there are mypy errors in the stubs.

With this patch:

<details>

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 46014bcd2..77e5e3127 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -166,3 +166,6 @@ known-first-party = ["ts_utils", "_utils"]
 
 [tool.typeshed]
 oldest_supported_python = "3.8"
+
+[tool.mypy]
+disable_error_code = "deprecated"
diff --git a/tests/stubtest_stdlib.py b/tests/stubtest_stdlib.py
index 5134f8270..77dab3d20 100755
--- a/tests/stubtest_stdlib.py
+++ b/tests/stubtest_stdlib.py
@@ -29,6 +29,8 @@ def run_stubtest(typeshed_dir: Path) -> int:
         "--check-typeshed",
         "--custom-typeshed-dir",
         str(typeshed_dir),
+        "--mypy-config-file",
+        str(typeshed_dir / "pyproject.toml"),
         *allowlist_stubtest_arguments("stdlib"),
     ]
     if sys.version_info < (3, 10):
```

</details>

I get these in the stdlib on macOS and 3.12:

```
error: asyncio.__all__ is not present in stub
error: email.__all__ is not present in stub
error: email._policybase.__all__ is not present in stub
error: importlib.resources.abc.__all__ is not present in stub
error: os.__all__ is not present in stub
error: socket.__all__ is not present in stub
error: xml.__all__ is not present in stub
error: zipfile._path.__all__ is not present in stub
note: unused allowlist entry concurrent.futures.__all__
```

Didn't run it on all the third-party stubs.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by srittau** [comment]
Interesting. I just commented out `__all__` in stubs/passlib/passlib/utils/__init__.pyi, ran stubtest on passlib and didn't get any warnings. There is `__all__` at runtime.

**Comment by JelleZijlstra** [comment]
Did you use `stubtest_third_party.py`? Looks like it always installs the mypy from our requirements.txt, so it wouldn't pick up local changes.

**Comment by srittau** [comment]
Yes, it seems that way. I added debug statements to the installed mypy and those are not printed either.

**Comment by JelleZijlstra** [comment]
Yes, if you want to test you'll have to edit requirements.txt or otherwise hack the script. I think the stdlib results are enough for me to believe that this works as expected, so I'm happy to merge this, but I'll wait a little bit in case Shantanu/Alex/someone else seeing this has more feedback.

**Comment by Avasam** [comment]
I do have a question: Do we want to report cases where `__all__` is missing in stubs, but __all__ at runtime is redundant? (ie: already re-exports all symbols). I don't feel strongly either way, I just want to make sure the decision is explicit.

**Comment by srittau** [comment]
> I do have a question: Do we want to report cases where `__all__` is missing in stubs, but `__all__` at runtime is redundant? (ie: re-exports all symbols anyway, so star-imports wouldn't be affected). I don't feel strongly either way, I just want to make sure the decision is explicit.

I would say yes. Stubs beings closer to the implementation has value in itself, instead of special casing specific cases. Also, it means that `x.__all__` type checks correctly. There might be other use cases that we don't know about yet.

File: mypy/stubtest.py [hunk 1]
```diff
@@ -348,6 +348,8 @@ def verify_mypyfile(
             # Only verify the contents of the stub's __all__
             # if the stub actually defines __all__
             yield from _verify_exported_names(object_path, stub, runtime_all_as_set)
+        else:
+            yield Error(object_path + ["__all__"], "is not present in stub", stub, runtime)
```
Comment (hauntsaninja) [inline mypy/stubtest.py:352]:
```suggestion
            yield Error(object_path + ["__all__"], "is not present in stub", MISSING, runtime)
```

File: mypy/test/teststubtest.py [hunk 2]
```diff
@@ -2435,11 +2435,6 @@ def test_output(self) -> None:
         assert output == expected
 
     def test_ignore_flags(self) -> None:
-        output = run_stubtest(
```
Comment (hauntsaninja) [inline mypy/test/teststubtest.py:2438]: this test case should still pass (see my suggested change above)

**Review by hauntsaninja — CHANGES_REQUESTED** [review]

**Review by hauntsaninja — APPROVED** [review]
Thank you!

**Pull request merged**